api/services/device.go

```diff
@@ -149,10 +164,39 @@ func (s *service) UpdatePendingStatus(ctx context.Context, uid models.UID, statu
 		return NewErrDeviceStatusAccepted(nil)
 	}
 
+	ns, err := s.store.NamespaceGet(ctx, tenant)
+	if err != nil {
+		return NewErrNamespaceNotFound(tenant, err)
+	}
+
 	if status != StatusAccepted {
 		return s.store.DeviceUpdateStatus(ctx, uid, status)
 	}
 
+	// NOTICE: The logic below is only executed when the new status is "accepted".
+
+	if ns.MaxDevices > 0 {
+		slot, err := s.store.DeviceRemovedGet(ctx, tenant, uid)
+		if err != nil && err != store.ErrNoDocuments {
+			return NewErrDeviceRemovedGet(err)
+		}
+
+		if slot == nil {
+			removed, err := s.store.DeviceRemovedList(ctx, tenant)
+			if err != nil {
+				return NewErrDeviceRemovedList(err)
+			}
+
+			if (ns.DevicesCount + len(removed)) >= ns.MaxDevices {
+				return NewErrDeviceRemovedFull(ns.MaxDevices, nil)
+			}
```

Comment:
Making the list of removed devices while a simple count using the tenant as a filter would do the work.